File: coupler.f90
!****************************************************************
!Toy coupled model to demonstrate MPI communicator concepts
!Author Ghazal Tashakor
!Email g.tashakor@fz-juelich.de
!****************************************************************

program Coupler
  implicit none
  include 'mpif.h'
  
  integer :: ierr, rank, nprocs, color=1, key=0
  integer :: PA_rank, PB_rank, PC_rank, PA_nprocs, PB_nprocs, PC_nprocs
  integer :: status(MPI_STATUS_SIZE)
  integer, dimension (:), allocatable :: PA_plist, PB_plist, PC_plist
  real(8), dimension (:,:), allocatable :: u_PC
  INTEGER :: cart_comm, row_comm, col_comm
  INTEGER :: num_rows, num_cols, cart_dims(2), cart_periodic(2)
  INTEGER :: cart_coords(2), row_rank, col_rank
  INTEGER :: ii ! Added declaration for 'ii'
  
  call MPI_INIT(ierr)
  call MPI_COMM_SIZE(MPI_COMM_WORLD,nprocs,ierr)
  call MPI_COMM_RANK(MPI_COMM_WORLD,rank,ierr)
  
  ! Determine dimensions of the 2D decomposition
  num_rows = 2
  num_cols = nprocs / num_rows
  
  ! Create 2D Cartesian communicator
  cart_dims(1) = num_rows
  cart_dims(2) = num_cols
  cart_periodic(1) = .FALSE.
  cart_periodic(2) = .FALSE.
  call MPI_CART_CREATE(MPI_COMM_WORLD, 2, cart_dims, cart_periodic, .TRUE., cart_comm, ierr)
  
  ! Get coordinates and ranks in the 2D grid
  call MPI_CART_COORDS(cart_comm, rank, 2, cart_coords, ierr)
  call MPI_CART_RANK(cart_comm, cart_coords, row_rank, ierr)
  call MPI_CART_RANK(cart_comm, [cart_coords(2), cart_coords(1)], col_rank, ierr)
  
  if (row_rank == 0 .and. col_rank == 0) then
      print *, 'Coupler : Global rank=', rank, 'Local rank=', row_rank, col_rank
  end if
  
  ! Determine the ranks of PA, PB, and PC components
  PA_rank = 0
  PB_rank = num_rows
  PC_rank = nprocs - 1
  
  if (row_rank == 0 .and. col_rank == 0) then
      print *, 'Coupler : PA_rank=', PA_rank
      print *, 'Coupler : PB_rank=', PB_rank
      print *, 'Coupler : PC_rank=', PC_rank
  end if
  
  ! Sending coupler global rank to PA, PB, and PC components
  call MPI_SEND(rank, 1, MPI_INT, PA_rank, 101, MPI_COMM_WORLD, ierr)
  call MPI_SEND(rank, 1, MPI_INT, PB_rank, 101, MPI_COMM_WORLD, ierr)
  call MPI_SEND(rank, 1, MPI_INT, PC_rank, 101, MPI_COMM_WORLD, ierr)
  
  ! Receiving num. procs and processor lists from PA, PB, and PC components
  if (row_rank == 0 .and. col_rank == 0) then
      call MPI_RECV(PA_nprocs, 1, MPI_INT, PA_rank, 102, MPI_COMM_WORLD, status, ierr)
      call MPI_RECV(PB_nprocs, 1, MPI_INT, PB_rank, 102, MPI_COMM_WORLD, status, ierr)
      call MPI_RECV(PC_nprocs, 1, MPI_INT, PC_rank, 102, MPI_COMM_WORLD, status, ierr)
      allocate(PA_plist(PA_nprocs))
      allocate(PB_plist(PB_nprocs))
      allocate(PC_plist(PC_nprocs))
  end if
  
  call MPI_GATHER(rank, 1, MPI_INT, PA_plist, 1, MPI_INT, PA_rank, row_comm, ierr)
  call MPI_GATHER(rank, 1, MPI_INT, PB_plist, 1, MPI_INT, PB_rank, row_comm, ierr)
  call MPI_GATHER(rank, 1, MPI_INT, PC_plist, 1, MPI_INT, PC_rank, row_comm, ierr)
  
  if (row_rank == 0 .and. col_rank == 0) then
      print *, 'Coupler : received PA processor list'
      print *, 'Coupler : received PB processor list'
      print *, 'Coupler : received PC processor list'
      call MPI_RECV(PA_plist, PA_nprocs, MPI_INT, PA_rank, 103, MPI_COMM_WORLD, status, ierr)
      call MPI_RECV(PB_plist, PB_nprocs, MPI_INT, PB_rank, 103, MPI_COMM_WORLD, status, ierr)
      call MPI_RECV(PC_plist, PC_nprocs, MPI_INT, PC_rank, 103, MPI_COMM_WORLD, status, ierr)
  end if
  
  call MPI_BARRIER(MPI_COMM_WORLD, ierr)
  
  if (row_rank == 0 .and. col_rank == 0) then
      ! Sending u_PC to PA component
      allocate(u_PC(190, 384))
      call random_seed()
      call random_number(u_PC)
      call MPI_SEND(u_PC, 190*384, MPI_DOUBLE_PRECISION, PA_rank, 109, MPI_COMM_WORLD, ierr)
      print *, 'Coupler : sending u_PC to PA'
  end if
  
  call MPI_BARRIER(MPI_COMM_WORLD, ierr)
  
  do
      if (row_rank == 0 .and. col_rank == 0) then
          call MPI_RECV(u_PC, 190*384, MPI_DOUBLE_PRECISION, PC_rank, 109, MPI_COMM_WORLD, status, ierr)
          print *, 'Coupler : received u_PC from PC'
          call MPI_SEND(u_PC, 190*384, MPI_DOUBLE_PRECISION, PB_rank, 109, MPI_COMM_WORLD, ierr)
          print *, 'Coupler : sending u_PC to PB'
      end if
      
      if (row_rank == 0 .and. col_rank == 0) then
          call MPI_RECV(u_PC, 190*384, MPI_DOUBLE_PRECISION, PB_rank, 113, MPI_COMM_WORLD, status, ierr)
          print *, 'Coupler : received u_PB from PB'
          call MPI_SEND(u_PC, 190*384, MPI_DOUBLE_PRECISION, PC_rank, 113, MPI_COMM_WORLD, ierr)
          print *, 'Coupler : sending u_PB to PC'
      end if
      
      call MPI_BARRIER(MPI_COMM_WORLD, ierr)
      ! Other operations and communication between components
      
      if (row_rank == 0 .and. col_rank == 0) then
          print *, 'Coupler doing work'
          call sleep(1)
          print *, 'Coupler done with work'
      end if
      
      call MPI_BARRIER(MPI_COMM_WORLD, ierr)
      
      if (row_rank == 0 .and. col_rank == 0) then
          ii = ii + 1
          print *, 'Coupler : current time is', ii*3600, 'seconds'
      end if
      
      if (row_rank == 0 .and. col_rank == 0) then
          if (ii == 24) exit
      end if
      
      call MPI_BARRIER(MPI_COMM_WORLD, ierr)
      
      if (row_rank == 0 .and. col_rank == 0) then
          ii = ii + 1
      end if
      
      call MPI_BARRIER(MPI_COMM_WORLD, ierr)
  end do
  
  call MPI_FINALIZE(ierr)
end program

File: Coupler.c
!****************************************************************
!Toy coupled model to demonstrate MPI communicator concepts
!Author Ghazal Tashakor
!Email g.tashakor@fz-juelich.de
!****************************************************************

#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

int main(int argc, char* argv[]) {
    int rank, nprocs, PA_rank, PB_rank, PC_rank, PA_nprocs, PB_nprocs, PC_nprocs;
    MPI_Status status;

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &nprocs);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    // Determine the ranks of PA, PB, and PC components
    PA_rank = 0;
    PB_rank = 1;
    PC_rank = 2;

    int num_iterations = 10; // Number of loop iterations

    for (int iteration = 0; iteration < num_iterations; iteration++) {
        if (rank == PA_rank) {
            // PA component

            // Receiving num. procs from coupler
            MPI_Recv(&PA_nprocs, 1, MPI_INT, PA_rank, 102, MPI_COMM_WORLD, &status);

            // Sending processor list to coupler
            int* PA_plist = (int*)malloc(PA_nprocs * sizeof(int));
            MPI_Gather(&rank, 1, MPI_INT, PA_plist, 1, MPI_INT, PA_rank, MPI_COMM_WORLD);
            MPI_Send(PA_plist, PA_nprocs, MPI_INT, PA_rank, 103, MPI_COMM_WORLD);

            // Receiving u_PC from coupler
            int u_PC_rows = 10; // Replace with the actual number of rows for u_PC
            int u_PC_cols = 10; // Replace with the actual number of columns for u_PC
            double** u_PC = (double**)malloc(u_PC_rows * sizeof(double*));
            for (int i = 0; i < u_PC_rows; i++) {
                u_PC[i] = (double*)malloc(u_PC_cols * sizeof(double));
            }
            MPI_Recv(&(u_PC[0][0]), u_PC_rows * u_PC_cols, MPI_DOUBLE, PA_rank, 109, MPI_COMM_WORLD, &status);

            // Perform PA computations
            printf("Iteration %d, Rank %d: Performing PA computations\n", iteration, rank);

            // Sending u_PA to coupler
            MPI_Send(&(u_PC[0][0]), u_PC_rows * u_PC_cols, MPI_DOUBLE, PA_rank, 110, MPI_COMM_WORLD);

            // Cleanup
            for (int i = 0; i < u_PC_rows; i++) {
                free(u_PC[i]);
            }
            free(u_PC);
            free(PA_plist);
        }
        else if (rank == PB_rank) {
            // PB component

            // Receiving num. procs from coupler
            MPI_Recv(&PB_nprocs, 1, MPI_INT, PB_rank, 102, MPI_COMM_WORLD, &status);

            // Sending processor list to coupler
            int* PB_plist = (int*)malloc(PB_nprocs * sizeof(int));
            MPI_Gather(&rank, 1, MPI_INT, PB_plist, 1, MPI_INT, PB_rank, MPI_COMM_WORLD);
            MPI_Send(PB_plist, PB_nprocs, MPI_INT, PB_rank, 103, MPI_COMM_WORLD);

            // Receiving u_PC from coupler
            int u_PC_rows = 10; // Replace with the actual number of rows for u_PC
            int u_PC_cols = 10; // Replace with the actual number of columns for u_PC
            double** u_PC = (double**)malloc(u_PC_rows * sizeof(double*));
            for (int i = 0; i < u_PC_rows; i++) {
                u_PC[i] = (double*)malloc(u_PC_cols * sizeof(double));
            }
            MPI_Recv(&(u_PC[0][0]), u_PC_rows * u_PC_cols, MPI_DOUBLE, PB_rank, 113, MPI_COMM_WORLD, &status);

            // Perform PB computations
            printf("Iteration %d, Rank %d: Performing PB computations\n", iteration, rank);

            // Sending u_PB to coupler
            MPI_Send(&(u_PC[0][0]), u_PC_rows * u_PC_cols, MPI_DOUBLE, PB_rank, 114, MPI_COMM_WORLD);

            // Cleanup
            for (int i = 0; i < u_PC_rows; i++) {
                free(u_PC[i]);
            }
            free(u_PC);
            free(PB_plist);
        }
        else if (rank == PC_rank) {
            // PC component

            // Receiving num. procs from coupler
            MPI_Recv(&PC_nprocs, 1, MPI_INT, PC_rank, 102, MPI_COMM_WORLD, &status);

            // Sending processor list to coupler
            int* PC_plist = (int*)malloc(PC_nprocs * sizeof(int));
            MPI_Gather(&rank, 1, MPI_INT, PC_plist, 1, MPI_INT, PC_rank, MPI_COMM_WORLD);
            MPI_Send(PC_plist, PC_nprocs, MPI_INT, PC_rank, 103, MPI_COMM_WORLD);

            // Receiving u_PA from PA component
            int u_PC_rows = 10; // Replace with the actual number of rows for u_PC
            int u_PC_cols = 10; // Replace with the actual number of columns for u_PC
            double** u_PC = (double**)malloc(u_PC_rows * sizeof(double*));
            for (int i = 0; i < u_PC_rows; i++) {
                u_PC[i] = (double*)malloc(u_PC_cols * sizeof(double));
            }
            MPI_Recv(&(u_PC[0][0]), u_PC_rows * u_PC_cols, MPI_DOUBLE, PC_plist[PA_rank], 110, MPI_COMM_WORLD, &status);

            // Perform PC computations
            printf("Iteration %d, Rank %d: Performing PC computations\n", iteration, rank);

            // Sending u_PC to coupler
            MPI_Send(&(u_PC[0][0]), u_PC_rows * u_PC_cols, MPI_DOUBLE, PC_rank, 111, MPI_COMM_WORLD);

            // Cleanup
            for (int i = 0; i < u_PC_rows; i++) {
                free(u_PC[i]);
            }
            free(u_PC);
            free(PC_plist);
        }
        else {
            // Coupler component

            // Sending num. procs to PA, PB, and PC components
            PA_nprocs = 1; // Set the number of procs for PA component
            PB_nprocs = 1; // Set the number of procs for PB component
            PC_nprocs = 1; // Set the number of procs for PC component

            MPI_Send(&PA_nprocs, 1, MPI_INT, PA_rank, 102, MPI_COMM_WORLD);
            MPI_Send(&PB_nprocs, 1, MPI_INT, PB_rank, 102, MPI_COMM_WORLD);
            MPI_Send(&PC_nprocs, 1, MPI_INT, PC_rank, 102, MPI_COMM_WORLD);

            // Receiving processor lists from PA, PB, and PC components
            int* PA_plist = (int*)malloc(PA_nprocs * sizeof(int));
            int* PB_plist = (int*)malloc(PB_nprocs * sizeof(int));
            int* PC_plist = (int*)malloc(PC_nprocs * sizeof(int));
            MPI_Gather(&rank, 1, MPI_INT, PA_plist, 1, MPI_INT, PA_rank, MPI_COMM_WORLD);
            MPI_Gather(&rank, 1, MPI_INT, PB_plist, 1, MPI_INT, PB_rank, MPI_COMM_WORLD);
            MPI_Gather(&rank, 1, MPI_INT, PC_plist, 1, MPI_INT, PC_rank, MPI_COMM_WORLD);
            MPI_Send(PA_plist, PA_nprocs, MPI_INT, PA_rank, 103, MPI_COMM_WORLD);
            MPI_Send(PB_plist, PB_nprocs, MPI_INT, PB_rank, 103, MPI_COMM_WORLD);
            MPI_Send(PC_plist, PC_nprocs, MPI_INT, PC_rank, 103, MPI_COMM_WORLD);

            // Receiving u_PA from PA component
            int u_PC_rows = 10; // Replace with the actual number of rows for u_PC
            int u_PC_cols = 10; // Replace with the actual number of columns for u_PC
            double** u_PC = (double**)malloc(u_PC_rows * sizeof(double*));
            for (int i = 0; i < u_PC_rows; i++) {
                u_PC[i] = (double*)malloc(u_PC_cols * sizeof(double));
            }
            MPI_Recv(&(u_PC[0][0]), u_PC_rows * u_PC_cols, MPI_DOUBLE, PA_rank, 109, MPI_COMM_WORLD, &status);

            // Sending u_PC to PC component
            MPI_Send(&(u_PC[0][0]), u_PC_rows * u_PC_cols, MPI_DOUBLE, PC_rank, 109, MPI_COMM_WORLD);

            // Receiving u_PB from PB component
            MPI_Recv(&(u_PC[0][0]), u_PC_rows * u_PC_cols, MPI_DOUBLE, PB_rank, 113, MPI_COMM_WORLD, &status);

            // Sending u_PC to PB component
            MPI_Send(&(u_PC[0][0]), u_PC_rows * u_PC_cols, MPI_DOUBLE, PB_rank, 113, MPI_COMM_WORLD);

            // Receiving u_PC from PC component
            MPI_Recv(&(u_PC[0][0]), u_PC_rows * u_PC_cols, MPI_DOUBLE, PC_rank, 111, MPI_COMM_WORLD, &status);

            // Cleanup
            for (int i = 0; i < u_PC_rows; i++) {
                free(u_PC[i]);
            }
            free(u_PC);
            free(PA_plist);
            free(PB_plist);
            free(PC_plist);
        }
    }

    MPI_Finalize();
    return 0;
}
